packages/core/core/src/TargetResolver.js

```diff
@@ -37,6 +43,362 @@ const DEFAULT_PRODUCTION_ENGINES = {
+      pkgContents,
+      keys.map(({key, type, message}) => ({
+        key: `/targets/${targetName}/${key}`,
+        type: type,
+        message
+      }))
+    );
+  }
+  throw new ThrowableDiagnostic({
+    diagnostic: {
+      message,
+      origin: '@parcel/core',
+      filePath: pkgPath || undefined,
+      language: 'json',
+      codeFrame:
+        highlight && pkgContents
+          ? {
+              code: pkgContents,
+              codeHighlights: highlight
+            }
+          : undefined,
+      hints: hints || undefined
+    }
+  });
+}
+
+function parseEngines(
+  targetName: ?string,
+  engines: mixed
+): Engines | typeof undefined {
+  if (engines === undefined) {
+    return;
+  } else if (engines && typeof engines === 'object') {
+    let result = {};
+    for (let key in engines) {
+      let value: mixed = engines[key];
+      if (key === 'browsers' && Array.isArray(value)) {
+        value = value.map(v => {
+          if (typeof v !== 'string') {
+            throw new Error(
+              `Invalid value in engines.browsers array ${
+                targetName ? `for target "${targetName}"` : ''
+              }: ${String(v)}`
+            );
+          }
+          return v;
+        });
+      } else if (typeof value !== 'string') {
+        throw new Error(
+          `Invalid value for engines.${key} ${
+            targetName ? `for target "${targetName}"` : ''
+          }: ${String(value)}`
+        );
+      }
+      result[key] = value;
+    }
+    return result;
+  } else {
+    throw new Error(
+      `Invalid engines ${
+        targetName ? `for target "${targetName}"` : ''
+      }: ${String(engines)}`
+    );
+  }
+}
+
+const LIST_OF_DESCRIPTOR_KEYS = [
+  'context',
+  'engines',
+  'includenodeModules',
+  'outputFormat',
+  'publicUrl',
+  'distDir',
+  'sourceMap',
+  'distDir',
+  'isLibrary'
+];
+function parseDescriptor(
+  targetName: string,
+  descriptor: mixed,
+  pkgPath: ?FilePath,
+  pkgContents: ?string
+): {|
+  context?: EnvironmentContext | typeof undefined,
+  engines?: Engines | typeof undefined,
+  includeNodeModules?: boolean | Array<PackageName> | typeof undefined,
+  outputFormat?: OutputFormat | typeof undefined,
+  publicUrl?: string | typeof undefined,
+  distDir?: FilePath | typeof undefined,
+  sourceMap?: TargetSourceMapOptions | typeof undefined,
+  distDir: FilePath | typeof undefined,
+  isLibrary: boolean | typeof undefined
+|} {
+  if (!(descriptor && typeof descriptor === 'object')) {
+    throw new Error(`Empty descriptor for target "${targetName}"`);
+  }
+
+  let {
+    context,
+    distDir,
+    engines: _engines,
+    includeNodeModules: _includeNodeModules,
+    isLibrary,
+    outputFormat,
+    publicUrl,
+    sourceMap: _sourceMap,
+    ...rest
+  } = descriptor;
+
+  if (Object.keys(rest).length !== 0) {
+    let keys = Object.keys(rest)
+      .map(k => {
+        let hint;
+        for (let l of LIST_OF_DESCRIPTOR_KEYS) {
```

Comment:
Should we use some fuzzy string matching algorithm to get the closest hint? 😜

Reply:
I'm now using all possible values sorted by their levenshtein distance that require less than `actualValue.length / 2` modifications. Otherwise all possible values are printed (and this distinction is also reflected in the hints).

![Bildschirmfoto 2019-11-01 um 14 09 38](https://user-images.githubusercontent.com/4586894/68026901-42b3c780-fcb1-11e9-8289-8a07fc02a693.png)

Performance-wise, in the screenshot these calculations take 0.879ms, 0.179ms, 0.063ms, 0.518ms. (Clearly correlated with the number of items).

This could probably be optimized with a levenshtein automaton instead (where the calculation aborts after the max number of steps is reached rather than filtering them out afterwards), but that would probably be overkill?